### ISSUE TITLE
Collecting Requirements for Per-Language Splitting

### ISSUE THREAD

**Issue by LorisSigrist** [opral/inlang-paraglide-js#88]
### Context

Paraglide currently splits messages by component / page. If you load a page with 3 client components (or your framework's equivalent) only the messages for those three components are sent to the client. But, they are currently sent in *all* languages. Ideally we would only send messages in the language that is displayed. 

This issue collects ideas on how that could be achieved. 

### Work done so far

We have already tried a few approaches & run into various challenges.

* Copying the `routes/` directory for each language & using middleware to multiplex between the different builds based on language.
  * Imports from in/out of the `routes/` folder are incredibly fragile
  * Doesn't work for all routers
  * Only works if the framework has a rewrite mechanism
* Post-processing the build output by copying each output file for each language and replacing messages with the language-specific version.
  * Doesn't work with compressed build ouputs
  * Introduces various linking issues

Fundamentally this is a *dynamic linking problem* in a world of ESM and *static linking*, which is really hard.

Another promising idea that we haven't tried *yet* is to serialize the messages & pass them along with the page-data. However, there are open questions on how we would know *which* messages need to be sent .

### Note: Lazy Loading is not the Solution

Any solution using `fetch` or `await import` is bound to introduce a render-fetch waterfall which drastically increases Time-To-Interactive. Eagerly loading messages in all languages is preferable in the vast majority of cases. 

Most projects have between 2-5 languages, lazy-loading only becomes justifiable at 10<.

**Comment by osdiab** [comment]
Keenly watching this. Seems like a core make or break feature that determines if this library can truly scale.

**Comment by osdiab** [comment]
> Another promising idea that we haven't tried yet is to serialize the messages & pass them along with the page-data. However, there are open questions on how we would know which messages need to be sent

Maybe leveraging AsyncLocalStorage (NextJS already seems to use this for headers()) to have a request context for this could help, having the translation functions add to a list at runtime?

**Comment by LorisSigrist** [comment]
That's an interesting idea, however, that likely only catches the messages that are actually executed during server-rendering, not messages that are used conditionally. We would need those too.

**Comment by osdiab** [comment]
Hmm yeah, in that case it probably can’t be a runtime thing then. Maybe can crawl the AST at compile time to find every reference to a translation function invoked in the starting point for each route (I think should be clear for each metaframework, eg for NextJS any default export from a page/layout/route file, not sure how one would achieve this framework agnostically though).

**Comment by minht11** [comment]
I think this could be solved with [import maps](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/script/type/importmap). It allows to load different specifiers dynamically. Main caveat being that importmap must be inserted before any module loading occurs. 

With this only one language would be loaded. The downside that i18n module keys could not be inlined inside the bundle like they are right now or whole application chunks would need to be duplicated.

I tested it locally and it works, I am pretty sure something like this could be implemented by Paraglide relatively easily.

Main experiment code, missing `en.js` and `de.js` files.
```html
<!doctype html>
<html lang="en">
  <head>
    <meta charset="UTF-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" />

    <script type="application/javascript">
      const language = localStorage.getItem('language') ?? 'en';

      const im = document.createElement('script');
      im.type = 'importmap';
      im.textContent = JSON.stringify({
        imports: {
          language: `/${language}.js`,
        }
      });
      document.currentScript.after(im);
    </script>
    
  </head>
  <body>
    <div id="languageContent"></div>

    <button id="toggleLanguage">
      Toggle language
    </button>

    <script type="module">
      import { languageName } from 'language'

      languageContent.innerHTML = `<h1>${languageName}</h1>`

      toggleLanguage.addEventListener('click', async () => {
        const language = localStorage.getItem('language') ?? 'en';
        const newLanguage = language === 'en' ? 'de' : 'en';

        localStorage.setItem('language', newLanguage);

        window.location.reload();
      });
    </script>
  </body>
</html>

```

**Comment by osdiab** [comment]
I think the problem with that is that the goal is to pass the i18n strings in the initial page load, not in a separate HTTP request after the page loads in the browser. The script tag would need to be parsed and executed in the client’s browser rather than happening entirely on the server.

**Comment by minht11** [comment]
If you inline script inside html, no seperate request will be made, it will load sync with html and since script for import map is very small the cost is very minimal, far less than even loading 2 languages.

Things won't be loaded lazy loaded, just i18n strings need to be in separate chunks for specifier imports to work (or whole separate app bundle for each language)

Also my solution would work for SPA too, in my use case I am not using a server or meta framework. Server solution you were discussing sounds like meta framework specific.

**Comment by LorisSigrist** [comment]
Inlining the scripts would be very nice! We'll definitely prototype that. <br>I'm not yet sure how we would do the build-transforms necessary to do this.

* How do we know which messages can be rendered on the current page? This info exists in the tree-shaking but we need to have it at runtime.
* How do we get the client-side build to use the inlined messages? 

It's a promising approach though. I imagine it would generalize quite well across frameworks

**Comment by ambigos1** [comment]
Hi @LorisSigrist .
I just noticed that all languages are getting downloaded and I found it after I translated my app to 57 languages :(

Performance is my top priority and deploying my website with 57 languages will damage my performance on web core vitals.

I am currently using **Svelte Static-Adapter** and all my website is prerendered for all languages.
Is there a way to prevent downloading the JavaScript files with all languages since they are generated on build time as static HTML files?

Thank you very much for your time :)

**Comment by samuelstroschein** [comment]
Hi @ambigos1 

> Is there a way to prevent downloading the JavaScript files with all languages since they are generated on build time as static HTML files?

Not atm. I might set a bounty on this issue. If anyone is down to implement per language-splitting after https://github.com/opral/inlang-paraglide-js/issues/217

**Comment by samuelstroschein** [comment]
The new vite environments https://main.vitejs.dev/guide/api-environment could be the solution we waited for by creating one environment per locale.

**Comment by ambigos1** [comment]
@samuelstroschein 
Vite 6 is out!
https://vite.dev/blog/announcing-vite6
:)

**Comment by ambigos1** [comment]
Now that the Vite 6 is out I hope Per Language will be done as soon as possible because now I am getting Core Web Vitals Assessment: **_Failed_**


![image](https://github.com/user-attachments/assets/c0e9a674-c9be-4fa3-92df-ec8737c27cf3)


> The new vite environments https://main.vitejs.dev/guide/api-environment could be the solution we waited for by creating one environment per locale.

**Comment by samuelstroschein** [comment]
@ambigos1 it seems like you are breaking tree-shaking somewhere and all messages are loaded. are you re-exporting messages from `i18n.js`? 

```ts
export * as m from "./paraglide/messages.js"
```

**Comment by ambigos1** [comment]
> @ambigos1 it seems like you are breaking tree-shaking somewhere and all messages are loaded. [paraglide js scales better than other i18n libs for <10 languages](https://inlang.com/m/gerre34r/library-inlang-paraglideJs/scaling). are you re-exporting messages from `i18n.js`?
> 
> ```ts
> export * as m from "./paraglide/messages.js"
> ```

I am using Manuel chunks with rollups this is why you see package called i18n, this is the name I gave it.

![image](https://github.com/user-attachments/assets/479b8272-86a6-447a-8b71-e458593c1bdd)


I am importing in the following way
	import * as m from '$lib/paraglide/messages.js';

and using it in JS files as
const title: string = m.main_page_title()

and on HTML section as follows:
{m.main_page_h2()}

**Comment by samuelstroschein** [comment]
@ambigos1 created a separate issue to not convoluted this one. i suspect that something is wrong with your setup, unrelated to language splitting.

**Comment by samuelstroschein** [comment]
Making the locale/language tag getter static on a per build basis could be interesting. If the language getter is static on a given build, bundlers will tree-shake unused imports. 

```diff
const jojo_mountain_day = (inputs, options = {}) => {
	const locale = "en";
	if (locale === "en") return en.jojo_mountain_day(inputs);
-	if (locale === "de") return de.jojo_mountain_day(inputs);
-	if (locale === "en-US") return en_US.jojo_mountain_day(inputs);
	return "jojo_mountain_day";
};
```

**Comment by samuelstroschein** [comment]
I will look into per language/locale splitting next week https://github.com/opral/inlang-paraglide-js/issues/201#issuecomment-2613623635

**Comment by samuelstroschein** [comment]
Trivial to implement for the compiler. Last open question is how to build per locale with bundlers. Vite's environment API could be the breakthrough. 

- add a `staticLocale` to the runtime
- define staticLocale on compile
- bundler tree-shakes messages that dont' correspond to the static locale 🚀  

```diff
+const staticLocale = "de"

const greeting = (inputs, options = {}) => {
+	const locale = staticLocale ?? options.locale ?? getLocale();
	if (locale === "en") return en.greeting(inputs);
	if (locale === "de") return de.greeting(inputs);
	return "greeting";
};

export { greeting };
```

**Comment by moufmouf** [comment]
Hey!

For the context, I'm in a situation where I have 10+ languages to translate and a SPA running in Svelte only mode (no SvelteKit). See #351.

I just had an idea I wanted to share here. I don't think it was mentioned before.

If I understand correctly, you are scratching your heads to avoid lazy-loading because you assume it will cause an additional round-trip:

> Any solution using fetch or await import is bound to introduce a render-fetch waterfall which drastically increases Time-To-Interactive. Eagerly loading messages in all languages is preferable in the vast majority of cases.

BUT! There are now ways to tell the browser to prefetch resources. For instance: Early hints: https://developer.mozilla.org/en-US/docs/Web/HTTP/Status/103

Basically, there might be no need to make one build per locale if you can ensure the correct message file is already loaded in the browser when your app is bootstrapping.

Did you guys already explore this solution?

**Comment by samuelstroschein** [comment]
@moufmouf thanks for the idea. might be something to it. 

The waterfall is not the main issue. The main issue is avoiding message functions like `m.happy_elephant()` to be async. The moment that happens, complexity will explode. Suspenses need to be implemented, waterfalls emerge, etc. 

What could work, however, is using ESM new top-level await. If the locale is set before the top level await of a message bundle function (the bundle function "bundles" the messages for all locales) is executed, then your approach could work! The bundler tree-shakes un-unsed message bundle functions, and the message bundle function lazy loads the message defined by the locale!

I will investigate this! This might have legs!

**Comment by ambigos1** [comment]
> Hey!
> 
> For the context, I'm in a situation where I have 10+ languages to translate and a SPA running in Svelte only mode (no SvelteKit). See [#351](https://github.com/opral/inlang-paraglide-js/issues/351).
> 
> I just had an idea I wanted to share here. I don't think it was mentioned before.
> 
> If I understand correctly, you are scratching your heads to avoid lazy-loading because you assume it will cause an additional round-trip:
> 
> > Any solution using fetch or await import is bound to introduce a render-fetch waterfall which drastically increases Time-To-Interactive. Eagerly loading messages in all languages is preferable in the vast majority of cases.
> 
> BUT! There are now ways to tell the browser to prefetch resources. For instance: Early hints: https://developer.mozilla.org/en-US/docs/Web/HTTP/Status/103
> 
> Basically, there might be no need to make one build per locale if you can ensure the correct message file is already loaded in the browser when your app is bootstrapping.
> 
> Did you guys already explore this solution?

I am using prerender, and all the text is generated at compile time.
I don't even need all the messages to be loaded at the production
If I only knew how I could remove them from the build output, it would help me boosting my performance

**Comment by samuelstroschein** [comment]
I am going to look into https://github.com/opral/inlang-paraglide-js/issues/88#issuecomment-2622529403 next week. 

In the meantime, I found https://github.com/tc39/proposal-defer-import-eval, which seems to be exactly what we need. Dereffered (sync!) imports that only load a message for a given locale if needed without async code. TC39 plans to advance the proposal to [Stage 3](https://tc39.es/process-document/) this month.

**Comment by samuelstroschein** [comment]
### Investigating top-level dynamic imports

Using dynamic top-level imports works, but there are edge cases that break. 

`await import(`./${getLocale()}.js`)` works as expected on the client IF the path is importable. The path is dynamic though, bundlers don't know what `getLocale()` could be and therefore create no importable chunk on the client (except if the dev server is running). 

```ts
// eslint-disable
import { getLocale, isServer } from '../../runtime.js';
import * as en from './en.js';
import * as de from './de.js';

let msg = undefined

if (!isServer){
	msg = await import(`./en.js`);
}
/* @__NO_SIDE_EFFECTS__ */
export const welcome = (inputs = {}, options = {}) => {
	if (!isServer) {
		// @ts-expect-error - xyz
		return msg.welcome(inputs);
	} else {
		const locale = options.locale ?? getLocale();
		if (locale === 'en') return en.welcome(inputs);
		if (locale === 'de') return de.welcome(inputs);
	}
	return 'welcome';
};
```

What if we place the conditional import logic to the top level scope as well? 

```ts
// eslint-disable
import { getLocale, isServer } from '../../runtime.js';

let msg = undefined

if (isServer === false){
	const locale = getLocale();
	if (locale === "en"){
		msg = await import('./en.js');
	} else if (locale === "de"){
		msg = await import('./de.js');
	}else {
		msg = { welcome: () => "welcome" }
	}
} else { 
	msg = {
		en: await import('./en.js'),
		de: await import('./de.js'),
	}
}


/* @__NO_SIDE_EFFECTS__ */
export const welcome = (inputs = {}, options = {}) => {
	if (isServer === false) {
		return msg?.welcome(inputs);
	} else {
		// @ts-expect-error - this is a server only function
		const locale = options.locale ?? getLocale();
		if (locale === 'en') return msg.en.welcome(inputs);
		if (locale === 'de') return msg.de.welcome(inputs);
		return 'welcome';
	}
};
```

This works 🥳 

The problem is that the client will now make N requests per message that is rendered. Making N requests could negatively impact metrics other than the bundle size. 


https://github.com/user-attachments/assets/b8d2b11c-56b2-4737-975c-ec300f3dc57a

 ## Next steps

- [ ] investigate impact if 200 message are rendered on first load page etc
- [ ] does this approach work in non SSR environments?
- [ ] does a way exist for the bundler to create one chunk per page dynamically

**Comment by samuelstroschein** [comment]
## Results of benching dynamic imports

As expected, dynamic imports are a de-optimization. Unexpecteively, dynamically loading each module leads to de-optimization in the bundle size as well. I assume that gzipping plays a role. Dynamic imports are individual modules that don't benefit from shared characters being transferred.  

- increase of networking loading time from 66ms to 770ms :x:
- increase of bundle size from 42kb (including app stuff) to 129kb :x:

## Next steps

- [ ] try vite environment API to create client side bundles per locale

**Comment by samuelstroschein** [comment]
Discussion about the vite environment api is here https://github.com/vitejs/vite/discussions/16358#discussioncomment-12356499

**Comment by samuelstroschein** [comment]
I got per locale splitting to work in SSR. See https://github.com/opral/inlang-paraglide-js/issues/425#issuecomment-2692351073. Even better, the requirements are clear now:


> - the client never splits the locales because that breaks SPA and MPA rendering modes
> - the server has to prune locales in the response (as it's the only certain place where SSR is used without effecting SPAs and MPAs). 
> 
> Furthermore, if https://github.com/opral/inlang-paraglide-js/issues/354 is implemented, the requirement for per-locale splitting will likely be severely reduced. The majority of SSR's sites would ship 0kb JS as the usage of dynamic messages that have inputs or multiple variants is rare.

**Comment by samuelstroschein** [comment]
A new tc39 proposal https://github.com/tc39-transfer/proposal-deferred-reexports might make locales lazy without the need for build optimizations.

**Comment by ambigos1** [comment]
Hi @samuelstroschein 👋 First, thank you for your work on this library!

Would it be possible to prioritize this feature? I'd even suggest considering a community funding campaign to help support development — I imagine others would be happy to contribute to something that meaningfully improves performance.

Thanks again for all your efforts!

**Comment by samuelstroschein** [comment]
@ambigos1 if someone makes a PR that is maintainable, i can release the feature. 

Prioritizing the feature from my side will not happen. we don't need the feature for ourselves atm, and paraglide js already has substantially smaller bundle sizes, even with 20 locales and no splitting implemented, than other i18n libs [benchmark](https://inlang.com/m/gerre34r/library-inlang-paraglideJs/benchmark) 

regardless of a community PR, i think bundlers (i.e. vite environment API) or JavaScript itself ( [deferred re-exports](https://github.com/tc39-transfer/proposal-deferred-reexports)) will make per locale relatively simple sometime this year

**Comment by VityaSchel** [comment]
Is it possible to retrieve the messages with client side rendering? Currently the "experimentalMiddlewareLocaleSplitting" option only supports SSR environments but I have a SPA app which runs entirely clientside and 24 locales for my app, so bundle size is close to 5 mb and includes every language.. Is there any workaround for the current implementation? I think I can use react's lazy loading and Vite will split every language into its own chunk but I'm not sure exactly how to implement it

**Comment by samuelstroschein** [comment]
@VityaSchel i responded in https://github.com/VityaSchel/lufin/issues/13#issuecomment-3237758117. 

TL;DR

Unlikely that Paraglide JS is the root cause. Your build likely has an issue. Please continue the conversation in the other issue.

**Comment by samuelstroschein** [comment]
https://github.com/VityaSchel/lufin/issues/13#issuecomment-3237758117

**Comment by samuelstroschein** [comment]
I linked the issue and my response https://github.com/VityaSchel/lufin/issues/13#issuecomment-3237758117. 

> The "won't fix because it does not cause significant problems to most users" doesn't seem relevant considering how popular paraglide became and how many users in this thread previously expressed frustration because of this issue ;)

I appreciate your feedback. I am open to a solution. So far everything has been a dead end. The complexity was not worth the effort, especially since per message tree-shaking even with loading all locales (if you have per page chunking). See the benchmarks https://inlang.com/m/gerre34r/library-inlang-paraglideJs/benchmark.

It's seems like the problem might solve itself with lazy exports https://github.com/tc39/proposal-deferred-reexports

**Comment by yeargun** [comment]
with static serving, in terms of information theory, will it even be possible? My guts telling me no.

nginx -> whats the curLang header/cookie for the request? {
thejsonpathweuserightnow/en.json
thejsonpathweuserightnow/fr.json
...
}

So basically, the library must be designed in a way to export seperate jsons that, with simple nginx rule every request will  getting proxied to language specific file.

**Comment by samuelstroschein** [comment]
Paraglide JS 2.9.0 will have a `experimentalStaticLocale` flag that you can test and give feedback on https://github.com/opral/paraglide-js/pull/591. 

The tree-shaking works excellent but it requires plugging a server in front of your app that routes to the build per locale. That is doable but we ideally find an API that doesn't require a separate server process from the app itself.

**Comment by samuelstroschein** [comment]
@rbozan no. strict splitting per locale. why would you want to mix locales on one page?

**Comment by rbozan** [comment]
> [@rbozan](https://github.com/rbozan) no. strict splitting per locale. why would you want to mix locales on one page?

Because I'm using Inlang / Paraglide for a language learning application, so for a few translations I do want to show dual translations for the user's preferred language and his/her target language

**Comment by samuelstroschein** [comment]
@rbozan as long as you dont run into bundle size issues, you shouldn't worry about per locale splitting (if you want to show multiple locales anyways)

**Comment by devcaeg** [comment]
@samuelstroschein

This library is gaining popularity as TanStack Start becomes more widely used, since it is recommended in the TanStack documentation.

That said, is the option released in January regarding this issue still considered “experimental”?

**Comment by samuelstroschein** [comment]
@devcaeg yes because it only works for SSG builds / I'm waiting for feedback where it breaks. generally though, the per translation tree-shaking is outperforming any other solution on the market irrespective of locale bundling. see https://inlang.com/m/gerre34r/library-inlang-paraglideJs/benchmark

**Comment by devcaeg** [comment]
@samuelstroschein I'm testing "experimentalMiddlewareLocaleSplitting" in TanStack Start, and in development mode, the page loads and displays the content, but then there's a flash and this error appears:

```ts
globalThis.__paraglide.ssr.auth_otp_title is not a function
```

Edit:
I realized that it's because the component only runs on the client side, never on the server side, so it doesn't know that the translation exists on the server side and never sends it to the client.

**Comment by michael-haberzettel** [comment]
Hello, 
I just tried `experimentalStaticLocale` with a Sveltekit application, but I think I don't understand how it works because my bundle file still remains with the same size (500kb on my side with brotli compression 😢 ).

My setup: 

- I have 13 languages in my application with approximately 1500 keys.
- In `vite.config.js`, I configured the Vite plugin for paraglide as follows:

   ```js
   paraglideVitePlugin({
            experimentalStaticLocale: "'fr'", // enforce locale to test it
            project: "./project.inlang",
            outdir: "./src/lib/paraglide",
            strategy: ["url", "cookie", "baseLocale"]
        }),
   ```
- I'm using SSR and CSR in my pages.
- `@inlang/paraglide-js@2.18.1`

When I'm loading the application, I'm always redirected to `/fr` but the bundle size of translations doesn't change when checking on my network tab.

**Comment by michael-haberzettel** [comment]
I'm not sure I've seen the idea I'm about to propose (if so, disregard my message).

To keep things simple, one solution would be to have a translation file for each locale when compiling with ViteJS using a hashcontent. For example:

```
- build/en.f26548fae354.json
- build/fr.8e334fa5e8b13.json
```

Then, in the HTML template file, load the corresponding locale **synchronously** with the middleware, which will replace a token like [this]( https://inlang.com/m/gerre34r/library-inlang-paraglideJs/sveltekit#add-the-paraglidemiddleware-to-srchooksserverts)

```html
<script type="text/javascript" src="%currentLocaleScript%"></script>
<script type="text/javascript">loadParaglideLocale()</script>
```

The script `build/en.f26548fae354.json` would expose a function `loadParaglideLocale()` to initialize paraglide properly on client side.

This approach can be useful because we can apply a strong cache on locales files.

**Comment by samuelstroschein** [comment]
@michael-haberzettel experimental static locale might not be working atm due to a regression in vite 8 https://github.com/opral/paraglide-js/issues/668. try vite 7. 

Regarding your idea with hashed translation files per locale, you likely loose per message tree-shaking which is worth more than per locale tree-shaking. The crux is to both tree-shake per locale and per message

**Comment by michael-haberzettel** [comment]
Thanks for your response, I will try with vite 7.

If I understand correctly, even without `experimentalStaticLocale`,  my huge file should be splitted by routes instead of one global chunk ?

**Comment by samuelstroschein** [comment]
> If I understand correctly, even without experimentalStaticLocale, my huge file should be splitted by routes instead of one global chunk ?

Yes. By component even.

**Comment by JosipVuk** [comment]
Hey @samuelstroschein ,

  Could you add a way to easily set `nonce` on the injected script?

  With `experimentalMiddlewareLocaleSplitting`, the inline script has no `nonce`, so under a strict CSP ('strict-dynamic') it's blocked and fails to run - breaking hydration.

**Comment by samuelstroschein** [comment]
@JosipVuk yes, make a PR. i'll merge it

**Comment by samuelstroschein** [comment]
Update: [[PR #718](https://github.com/opral/paraglide-js/pull/718)](https://github.com/opral/paraglide-js/pull/718) implements compiler-based per-locale client builds for TanStack Start and SvelteKit.

## Expected impact

Paraglide already tree-shakes unused messages per component/page. Per-locale splitting adds another step: from the messages that remain, it removes every translation except the active locale.

With `L` similarly sized locales, the approximate reduction in client-side translation code is:

- 2 locales: ~50%
- 4 locales: ~75%
- 7 locales: ~86%
- 10 locales: ~90%

This applies to the translation portion, not the entire application bundle:

```text
estimated savings =
current client translation bytes × (locales - 1) / locales
```

Using the original Inlang.com example from this issue, approximately 8 KB of translations across seven locales becomes roughly 1.1 KB—about 7 KB saved per page.

The implementation does not lazy-load individual messages and introduces no render/fetch waterfall. Vite builds the normal client graph, then Paraglide specializes it at build time using Vite 8’s Oxc optimizer. Each visitor receives only the graph for their locale.

There are trade-offs:

- Build time and deployed asset storage grow roughly with the number of locales.
- The server bundle remains shared.
- Changing locale requires a full navigation and loads the other locale’s client graph.
- The total page-load improvement depends on how much of the page’s JavaScript is translation code. A large application with only 8 KB of translations saves approximately 7 KB, not 86% of all JavaScript.

## Usage

After #718 is merged and released:

```ts
paraglideVitePlugin({
  project: "./project.inlang",
  outdir: "./src/paraglide",
  experimentalPerLocaleBuild: true,
})
```

Paraglide reads the locales from the inlang project automatically and emits a locale-to-assets manifest.

For TanStack Start, re-export the generated server entry:

```ts
// src/server.ts
export { default } from "./paraglide/tanstack-start.server.js"
```

For SvelteKit, keep using the normal `paraglideMiddleware` hook. Add `data-sveltekit-reload` to links that switch locale.

`setLocale()` already performs a full navigation by default.

The initial experiment deliberately supports a narrow configuration:

- Vite 8
- TanStack Start 1.168.x or SvelteKit 2.69.x
- Root/default asset hosting
- A locale strategy beginning with `url` or `cookie`
- SvelteKit SSR or fully prerendered pages, not SPA fallback mode

Unsupported configurations fail during the build instead of silently producing mixed-locale assets.

This is deliberately experimental. If you try it and find an issue or unsupported configuration, please open a PR with a fix.

**Comment by CanRau** [comment]
Would `experimentalPerLocaleBuild` already ensure `baseLocale` fallback is included as well by default or as an additional opt-in/out?

So that `baseLocale` is always included as well as fallback for untranslated strings
Or would the build fail/error/warn for each missing message?
